feat: add a hierarchical Prometheus MetricsRegistry trait for DistributedRuntime, Namespace, Components, and Endpoint

keivenchang opened the pull request:
#### Overview:

Added hierarchical Metrics trait for DistributedRuntime, Namespace, Component, and Endpoint.

#### Details:

* Introduced a hierarchical metrics registry framework for Prometheus metrics, enabling auto labels in Prometheus for namespaces, components, and endpoints.
* Added a new example project, "system_metrics", demonstrating custom stats endpoints and metrics collection.
* Implemented MetricsRegistry struct, and provided trait implementations for drt, Namespace, Component, and Endpoint.

#### Where should the reviewer start?

lib/runtime/src/metrics.rs
lib/runtime/examples/system_metrics/src/bin/system_server.rs

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

DIS-216

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Introduced a hierarchical metrics registry framework for Prometheus integration, enabling unified metric management across distributed runtime, namespaces, components, and endpoints.
  * Added a new example project, "system_stats_endpoint," demonstrating custom stats endpoints and metrics collection.
  * Implemented new public structs and traits for metrics, including `MetricsRegistry`, and provided trait implementations for core entities.

* **Improvements**
  * Refactored HTTP server metrics handling to use the new metrics registry, with improved uptime tracking and Prometheus output.
  * Enhanced configuration and modularity for metrics and profiling in the distributed runtime.

* **Documentation**
  * Added a README for the new "system_metrics" example.

* **Tests**
  * Added comprehensive tests for hierarchical metric registration, uptime initialization, and Prometheus output formatting.

* **Chores**
  * Updated workspace configuration to include new examples and dependencies.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

keivenchang committed:
feat: adding a common Metrics API for metrics endpoints DIS-216

keivenchang committed:
added hierarchical registry

keivenchang committed:
Moved metrics registry from DistributedRuntime to Namespace - each namespace now has its own metrics registry

Others:
Added NamespaceMetrics struct for namespace-level metrics management
Refactored HTTP server - removed RuntimeMetrics, simplified to HttpServerState with direct registry access
Updated metric naming - changed from single underscore (test_metric) to double underscore (test__metric) for Prometheus compatibility
Enhanced hierarchical registries - improved parent-child registry relationships with create_child_registry() method
Updated system stats example - aligned component/endpoint names and registry usage patterns

keivenchang committed:
clean up the code

keivenchang committed:
more cleanup

keivenchang committed:
more fixes

ryanolson committed:
paired programming, Ryan w/Keiven for the Prometheus Registry

keivenchang committed:
version that compiles and passes tests

keivenchang committed:
completely working MetricsRegistry in Runtime and Components

keivenchang committed:
fulling working MetricsRegistry

keivenchang committed:
cargo fmt

rmccorm4 commented on lib/runtime/src/http_server.rs:
```diff
@@ -223,62 +247,86 @@ mod tests {
+    }
+
+    #[tokio::test]
+    async fn test_runtime_metrics_namespace() {
```
How does `test_runtime_metrics_creation` differ from `test_runtime_metrics_namespace`? Seems like tests are same or very similar at a glance - validating uptime metric with a hard-coded value 123.456 / 42.0.

rmccorm4 commented on lib/runtime/src/profiling.rs:
```diff
@@ -0,0 +1,700 @@
+        );
+        assert_eq!(endpoint.parent_hierarchy(), vec!["", "mynamespace", "mynamespace_mycomponent"], "Endpoint parent hierarchy should be [\"\", \"mynamespace\", \"mynamespace_mycomponent\"]");
+        assert_eq!(
+            endpoint.prefix(),
+            "mynamespace_mycomponent_myendpoint",
```
CC @biswapanda

keivenchang committed:
Use basename instead of prefix

keivenchang commented on lib/runtime/src/http_server.rs:
```diff
@@ -13,77 +13,102 @@
+            .map_err(|_| "Start time already initialized")
+    }
+
+    pub fn uptime(&self) -> std::time::Duration {
+        self.start_time
+            .get()
+            .expect("Start time not initialized")
+            .elapsed()
+    }
```
Done.

keivenchang committed:
updated such that there is no prefix in the metric name, and now with hierarchical labels

keivenchang committed:
modified profiling.rs to hide private functions

nnshah1 commented on lib/runtime/src/profiling.rs:
question - why not metrics_registry as a name? or metrics ? - profiling seems more about timing of things and not generic metrics?

nnshah1 replied:
I think we can have an additional file for any tracing - right now we have logging.rs that also includes tracing as well - so likely this should remain metrics

keivenchang replied:
Done, profiling.rs => metrics.rs

keivenchang replied:
Renamed to metrics.rs

keivenchang commented:
Addressed the concern-- I removed the hierarchical prefix, and added auto-labeling. Currently working addressing the nitpick items, but do please start the review while I fix them. Thanks.

nnshah1 commented on lib/runtime/examples/system_stats_endpoint/src/bin/system_stats_client.rs:
```diff
@@ -0,0 +1,54 @@
+        println!("{:?}", resp);
+    }
+
+    let service_set = component.scrape_stats(Duration::from_millis(100)).await?;
```
Question: should we rename this example something like "system_metrics" ? and should we remove scape_stats here - or is that all hooked up to the same logic now?  Just double checking - wasn't sure if the stats_handler is implemented via metrics now.

keivenchang replied:
The scrape_stats hits the stats_handler. This usage comes from service_client.rs and service_server.rs, to demonstrate how these two work. I have TODO item to turn stats_handler to handle the Prometheus outputs: https://linear.app/nvidia-dynamo/issue/DIS-281/rust-stats-handler-to-publish-from-metricsregistry-to-json

keivenchang committed:
renaming from profiling.rs to metrics.rs, and the example metrics server/client

keivenchang committed:
add README in the example and fix nitpicks

keivenchang committed:
very minor spacing reverts, and adding comment

keivenchang committed:
cargo fmt

keivenchang committed:
fix vec macros

keivenchang committed:
fix unused var error

keivenchang committed:
add ignore NATS during testing

keivenchang committed:
Updated the create_test_drt() methods

keivenchang committed:
cargo fmt

keivenchang committed:
namespace to be unique in tests

keivenchang committed:
Set etcd/NATS tests to be integration

ryanolson reviewed (approved):
👍
